providers/Spotify/mod.ts

```diff
@@ -0,0 +1,321 @@
+import { ApiAccessToken, type CacheEntry, MetadataApiProvider, ReleaseApiLookup } from '@/providers/base.ts';
+import { DurationPrecision, FeatureQuality, FeatureQualityMap } from '@/providers/features.ts';
+import { parseHyphenatedDate, PartialDate } from '@/utils/date.ts';
+import { splitLabels } from '@/utils/label.ts';
+import { ResponseError } from '@/utils/errors.ts';
+import { selectLargestImage } from '@/utils/image.ts';
+import { encodeBase64 } from 'std/encoding/base64.ts';
+import { availableRegions } from './regions.ts';
+
+import type {
+	Album,
+	ApiError,
+	Copyright,
+	ResultList,
+	SearchResult,
+	SimplifiedArtist,
+	SimplifiedTrack,
+	Track,
+	TrackList,
+} from './api_types.ts';
+import type { ArtistCreditName, EntityId, HarmonyMedium, HarmonyRelease, HarmonyTrack } from '@/harmonizer/types.ts';
+
+// See https://developer.spotify.com/documentation/web-api
+
+const spotifyClientId = Deno.env.get('HARMONY_SPOTIFY_CLIENT_ID') || '';
+const spotifyClientSecret = Deno.env.get('HARMONY_SPOTIFY_CLIENT_SECRET') || '';
+
+export default class SpotifyProvider extends MetadataApiProvider {
+	readonly name = 'Spotify';
+
+	readonly supportedUrls = new URLPattern({
+		hostname: 'open.spotify.com',
+		pathname: '{/intl-:region}?/:type(artist|album)/:id',
+	});
+
+	readonly features: FeatureQualityMap = {
+		'cover size': 640,
+		'duration precision': DurationPrecision.MS,
+		'GTIN lookup': FeatureQuality.GOOD,
+		'MBID resolving': FeatureQuality.GOOD,
+		'release label': FeatureQuality.PRESENT,
+	};
+
+	readonly entityTypeMap = {
+		artist: 'artist',
+		release: 'album',
+	};
+
+	readonly availableRegions = new Set(availableRegions);
+
+	readonly releaseLookup = SpotifyReleaseLookup;
+
+	readonly launchDate: PartialDate = {
+		year: 2008,
+		month: 10,
+	};
+
+	readonly apiBaseUrl = 'https://api.spotify.com/v1/';
+
+	constructUrl(entity: EntityId): URL {
+		return new URL([entity.type, entity.id].join('/'), 'https://open.spotify.com');
+	}
+
+	async query<Data>(apiUrl: URL, maxTimestamp?: number): Promise<CacheEntry<Data>> {
+		try {
+			const accessToken = await this.cachedAccessToken(this.requestAccessToken);
+			const cacheEntry = await this.fetchJSON<Data>(apiUrl, {
+				policy: { maxTimestamp },
+				requestInit: {
+					headers: {
+						'Authorization': `Bearer ${accessToken}`,
+					},
+				},
+			});
+			const apiError = cacheEntry.content as ApiError;
+			if (apiError.error) {
+				throw new SpotifyResponseError(apiError, apiUrl);
+			}
+			return cacheEntry;
+		} catch (error) {
+			// Clone the response so the body of the original response can be
+			// consumed later if the error gets re-thrown.
+			const apiError = await error?.response?.clone().json() as ApiError;
```

Comment:
```suggestion
			// Clone the response so the body of the original response can be
			// consumed later if the error gets re-thrown.
			const apiError = await error.response?.clone().json() as ApiError;
```
Good thought, I always forget to do this before I encounter the issue with the already consumed body later on. `error` shouldn't be nullish though.